src/nbl/video/CVulkanBuffer.h

```diff
@@ -0,0 +1,55 @@
+#ifndef __NBL_C_VULKAN_BUFFER_H_INCLUDED__
+
+#include "nbl/video/IGPUBuffer.h"
+
+#include <volk.h>
+
+namespace nbl::video
+{
+
+class ILogicalDevice;
+
+class CVulkanBuffer : public IGPUBuffer
+{
+public:
+    CVulkanBuffer(core::smart_refctd_ptr<ILogicalDevice>&& dev, const IDriverMemoryBacked::SDriverMemoryRequirements& reqs, 
+        VkBuffer buffer) : IGPUBuffer(std::move(dev), reqs), m_vkBuffer(buffer)
+    {}
+
+    ~CVulkanBuffer();
+
+    inline VkBuffer getInternalObject() const { return m_vkBuffer; };
+
+    // Todo(achal): I don't think its possible
+    bool canUpdateSubRange() const override { return false; }
+
+    // Todo(achal)
+    //! Returns the allocation which is bound to the resource
+    IDriverMemoryAllocation* getBoundMemory() override
+    {
+        return nullptr;
+    }
+
+    // Todo(achal)
+    //! Constant version
+    const IDriverMemoryAllocation* getBoundMemory() const override
+    {
+        return nullptr;
+    }
```

Comment:
add a `core::smart_refctd_ptr<CVulkanMemoryAllocation> m_memory; uint64_t m_memBindingOffset;` to the privates

Reply:
When you create a `CVulkanBuffer` you don't have a handle to its memory (because it hasn't been allocated yet), so I would need a setter for setting `m_memory` (and probably `m_memBindingOffset`), which means `IGPUBuffer` should have a
`core::smart_refctd_ptr<IDriverMemoryAllocation>` private member (and probably same for `m_memBindingOffset`).

Am I allowed to make that change, or is there another way that I'm missing?

Reply:
Don't add the method to `IGPUBuffer`, also the members cannot be added to `IGPUBuffer` as that would result in a circular reference on OpenGL (`COpenGLBuffer` and `COpenGLImage` are both `IDriverMemoryBacked` and `IDriverMemoryAllocation`)

Only, `CVulkanBuffer` should have that method and the members.

Reply:
resolved

Reply:
Is it required that I put a `core::smart_refctd_ptr<CVulkanDriverMemoryAllocation>`?
If I do so I would have to `reinterpret_cast` for the getters, like:
```C
IDriverMemoryAllocation* getBoundMemory() override
{
    return reinterpret_cast<IDriverMemoryAllocation*>(m_memory.get());
}
```
which is not the case if I use a `core::smart_refctd_ptr<IDriverMemoryAllocation>`.

As I understand it, the sole reason to have this member here is to make sure that buffer's memory lives as long as the buffer, so it shouldn't matter if we hold a back pointer to `IDriverMemoryAllocation` or `CVulkanMemoryAllocation`?